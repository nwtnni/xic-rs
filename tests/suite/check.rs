use std::fmt;
use std::path::Path;

struct Snapshot<T>(Result<T, xic::Error>);

impl<T> fmt::Display for Snapshot<T> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        match &self.0 {
            Ok(_) => write!(fmt, "Valid Xi Program"),
            Err(error) => match error.report() {
                None => write!(fmt, "{}", error),
                Some(report) => {
                    let cache = xic::data::span::FileCache::default();
                    let mut buffer = Vec::new();

                    report
                        .with_config(ariadne::Config::default().with_color(false))
                        .finish()
                        .write(cache, &mut buffer)
                        .unwrap();

                    write!(fmt, "{}", String::from_utf8(buffer).unwrap())
                }
            },
        }
    }
}

#[test_generator::test_resources("tests/check/*.xi")]
pub fn check(path: &str) -> anyhow::Result<()> {
    let program = super::parse(path)?;
    let checked = xic::api::check(None, Path::new(path), program);

    insta::assert_display_snapshot!(path, Snapshot(checked));
    Ok(())
}
