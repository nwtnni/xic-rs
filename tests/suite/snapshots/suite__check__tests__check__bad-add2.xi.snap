---
source: tests/suite/check.rs
assertion_line: 35
expression: Snapshot(context)
---
Error: Semantic error
   ╭─[tests/check/bad-add2.xi:2:17]
   │
 2 │     x:int = 1 + {1,2}
   ·                 ──┬──  
   ·                   ╰──── Expected int but found int[]
───╯

